cdc/sinkv2/eventsink/cloudstorage/cloud_storage_dml_sink.go

```diff
@@ -0,0 +1,184 @@
+	model.TableName
+	TableVersion uint64
+}
+
+// eventFragment is used to attach a sequence number to TxnCallbackableEvent.
+// The sequence number is mainly useful for TxnCallbackableEvent defragmentation.
+// e.g. TxnCallbackableEvent 1~5 are dispatched to a group of encoding workers, but the
+// encoding completion time varies. Let's say the final completion sequence are 1,3,2,5,4,
+// we can use the sequence numbers to do defragmentation so that the events can arrive
+// at dmlWorker sequentially.
+type eventFragment struct {
+	// event sequence number
+	seqNumber    uint64
+	tableVersion uint64
+	tableName    model.TableName
+	event        *eventsink.TxnCallbackableEvent
+	// encodedMsgs denote the encoded messages after the event is handled in encodingWorker.
+	encodedMsgs []*common.Message
+}
+
+func eventFragmentLess(e1, e2 eventFragment) bool {
+	return e1.seqNumber < e2.seqNumber
+}
+
+// sink is the cloud storage sink.
+// It will send the events to cloud storage systems.
+type sink struct {
+	// msgChan is a unbounded channel to hold eventFragment.
+	msgChan *chann.Chann[eventFragment]
+	// encodingWorkers defines a group of workers for encoding events.
+	encodingWorkers []*encodingWorker
+	// writer is a dmlWriter which manages a group of dmlWorkers and
+	// sends encoded messages to individual dmlWorkers.
+	writer *dmlWriter
+	// tableSeqMap maintains a <versionTable, sequenceNumber> mapping.
+	tableSeqMap map[versionedTable]uint64
+	mu          sync.Mutex
+}
+
+// NewCloudStorageSink creates a cloud storage sink.
+func NewCloudStorageSink(ctx context.Context,
+	sinkURI *url.URL,
+	replicaConfig *config.ReplicaConfig,
+	errCh chan error,
+) (*sink, error) {
+	s := &sink{}
+	// create cloud storage config and then apply the params of sinkURI to it.
+	cfg := cloudstorage.NewConfig()
+	err := cfg.Apply(ctx, sinkURI, replicaConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	// parse backend storage from sinkURI.
+	bs, err := storage.ParseBackend(sinkURI.String(), nil)
+	if err != nil {
+		return nil, err
+	}
+
+	// create an external storage.
+	storage, err := storage.New(ctx, bs, &storage.ExternalStorageOptions{
+		SendCredentials: false,
+		S3Retryer:       rcommon.DefaultS3Retryer(),
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	// fetch protocol from replicaConfig defined by changefeed config file.
+	protocol, err := util.GetProtocol(replicaConfig.Sink.Protocol)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	// get cloud storage file extension according to the specific protocol.
+	ext := util.GetFileExtension(protocol)
+	encoderConfig, err := util.GetEncoderConfig(sinkURI, protocol, replicaConfig, defaultMaxMessageBytes)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	encoderBuilder, err := builder.NewEventBatchEncoderBuilder(ctx, encoderConfig)
+	if err != nil {
+		return nil, cerror.WrapError(cerror.ErrCloudStorageInvalidConfig, err)
+	}
+
+	changefeedID := contextutil.ChangefeedIDFromCtx(ctx)
```

Comment:
It's better pass the `changefeedID` explicitly by passing as argument instead of context.Value.

Reply:
For mq and txn sink in sinkV2, `changefeedID` are also extracted from `ctx`. Maybe in later PR, we can modify it at those places together.